funsor/distributions.py

```diff
@@ -113,11 +114,14 @@ def __repr__(self):
         return '{}({})'.format(type(self).__name__,
                                ', '.join('{}={}'.format(*kv) for kv in self.params.items()))
 
-    def __call__(self, *args, **kwargs):
-        kwargs = {d: to_funsor(v) for d, v in kwargs.items() if d in self.dims}
-        kwargs.update(zip(self.dims, map(to_funsor, args)))
-        if not kwargs:
-            return self
+    def _eager_subs(self, **kwargs):
+        # DEBUG
```

Comment:
@eb8680 I'm having some trouble using the `@Eager.register(Substitution)` decorator. I've added some print statements and the following code fails with:
```
$ pytest -vx test/ -k normal_density --pdb
...
DEBUG Funsor.__call__ Normal(loc=-2.25884914398, scale=0.941187381744, value=value) (Tensor((), 1.13507556915),) {}
DEBUG _eager_subs Normal(loc=-2.25884914398, scale=0.941187381744, value=value) (('value', Tensor((), 1.13507556915)),)
DEBUG Distribution._eager_subs Normal(loc=-2.25884914398, scale=0.941187381744, value=value) {'value': Tensor((), 1.13507556915)}
DEBUG Funsor.__call__ value () {'value': Tensor((), 1.13507556915)}
FunsorMeta._ast_call <class 'funsor.terms.Substitution'> (Variable('value', 'real'), (('value', Tensor((), 1.13507556915)),))
```
I don't understand why `Variable._eager_subs()` is not being called.

Reply:
I modified `Handler` so that, by default, effectful operations that happen inside a `Handler.process` are considered out of scope for that handler and are processed only by the handlers outside it on the stack.  In this case `var(**kwargs)` calls `Substitution`, which is considered out of scope by `Eager` because it happens inside `Eager.process`.  Here you would need to call `Variable._eager_subs` directly within `Distribution._eager_subs`.

The motivation for this experimental change was so that effects could be fused - for example, in `opteinsum_engine` the optimized interpreter is implemented as a single base interpreter pass via `EagerEval(Optimize(Deoptimize(Desugar(eval))))`

Reply:
Hmm, I'll have to think about this more. I guess I'll have to change e.g.
```diff
  def _eager_subs(self, **kwargs):
-     return Binary(self.op, self.lhs(**kwargs), self.rhs(**kwargs))
+     return eager_binary(self.op, self.lhs._eager_subs(**kwargs), self.rhs._eager_subs(**kwargs))
```

Reply:
Hmm I think the new non-reentrant behavior will prevent eager evaluation of terms like
```py
(Variable('x', 'real') + 1 + 1 + 1)(x=1)
```

Reply:
Ok, I'm making those changes now.